runtime/compiler/optimizer/HotFieldMarking.cpp

```diff
@@ -0,0 +1,173 @@
+                  {
+                  itr->second->_score = getReducedFrequencySum(itr->second->_score, block->getFrequency());
+                  } 
+               else
+                  {
+                  itr->second->_score = getReducedFrequencyAverage(itr->second->_score, itr->second->_count, block->getFrequency());
+                  itr->second->_count += 1;
+                  }
+               continue;
+               }
+
+            TR::ResolvedMethodSymbol *rms = comp()->getOwningMethodSymbol(symRef->getOwningMethodIndex());
+            TR_ResolvedMethod *method = rms->getResolvedMethod();
+
+            int32_t fieldNameLength = 0;
+            char *fieldName = method->fieldNameChars(symRef->getCPIndex(), fieldNameLength);
+            int32_t fieldSigLength = 0;
+            char *fieldSig = method->fieldSignatureChars(symRef->getCPIndex(), fieldSigLength);
+            bool isStatic = false;
+            TR_OpaqueClassBlock *containingClass = static_cast<TR_ResolvedJ9Method*>(method)->definingClassFromCPFieldRef(comp(), symRef->getCPIndex(), isStatic);
+            if (isStatic)
+               continue;
+
+            stats[symRef->getSymbol()] = new (trStackMemory()) SymStats(1, block->getFrequency(), fieldNameLength, fieldName, fieldSigLength, fieldSig, containingClass);
+            }
+         }
+      }
+
+   for (auto itr = stats.begin(), end = stats.end(); itr != end; ++itr)
+      {
+      if (itr->second->_score >= TR::Options::_hotFieldThreshold)
+         {
+         int32_t fieldOffset = (comp()->fej9()->getInstanceFieldOffset(itr->second->_clazz, itr->second->_fieldName, itr->second->_fieldNameLength, itr->second->_fieldSig, itr->second->_fieldSigLength) + TR::Compiler->om.objectHeaderSizeInBytes()) / TR::Compiler->om.sizeofReferenceField();
+            
+         if (!comp()->fej9()->isAnonymousClass(itr->second->_clazz))
+            {
+            comp()->fej9()->reportHotField(getUtilization(), TR::Compiler->cls.convertClassOffsetToClassPtr(itr->second->_clazz), fieldOffset, itr->second->_score);
+            
+            if (comp()->getOption(TR_TraceMarkingOfHotFields))
+               {
+               int32_t classNameLength = 0;
+               char *className = comp()->fej9()->getClassNameChars(itr->second->_clazz, classNameLength);
+               //dumpOptDetails(comp(), "HFM: %s notifying hot field %.*s%.*s.%.*s = %d\n", itr->second->_fieldSigLength, itr->second->_fieldSig, classNameLength, className, itr->second->_fieldNameLength, itr->second->_fieldName, itr->second->_score);           
+               printf("HFM: signature: %.*s; class:%.*s; fieldName: %.*s; frequencyScore = %d; fieldOffset1: %d; \n", J9UTF8_LENGTH(itr->second->_fieldSig), J9UTF8_DATA(itr->second->_fieldSig), J9UTF8_LENGTH(className), J9UTF8_DATA(className), J9UTF8_LENGTH(itr->second->_fieldName), J9UTF8_DATA(itr->second->_fieldName), itr->second->_score, fieldOffset);
```

Comment:
there shouldn't be any printf's in the JIT code, please use the tracing framework

Reply:
Ok, is the commented "dumpOptDetails" above the printf sufficient or should I just change the printf to use the traceMsg framework?

Reply:
because this is a block of trace guarded code you should use traceMsg - dumpOptDetails is used when the compiler is performing a transformation that is not optional - this is just tracing code.

Each opt gets its own trace option which you can test for enabled with trace(). Unless there is a reason for a different flag the code should use trace()